synapse/replication/tcp/streams/events.py

```diff
@@ -174,6 +174,10 @@ async def _update_function(
             upper_limit = state_rows[-1][0]
             limited = True
 
+            # FIXME: is it a given that there is only one row per stream_id in the
+            # state_deltas table (so that we can be sure that we have got all of the
+            # rows for upper_limit)?
```

Comment:
No it isn't a given.

Reply:
i.e. they can have multiple rows per stream ID afaict

Reply:
Yeah, I suspected as much. I was rather hoping that this wouldn't be any worse than it has been for years, and that it would therefore be safe to deploy while I thought about it a bit harder.